toy_2d/src/ilqr.py

```diff
@@ -0,0 +1,333 @@
+"""This file tries to solve the unconstrained optimization problem using the iLQR method
+The cost we want to minimize is
+
+           min         sum_{k=0}^{N-1} x_k^T Q x_k + u_k^T R u_k + x_N^T Q_N x_N
+           u_k
+
+        such that   x_{k+1} = f(x,u) + f(x)*lambda
+
+Given an initial guess, the problem first performs a rollout and then uses the backward pass to calculate the new controls by linearizing
+the dynamics and solving the riccati equation. 
+
+During the backward pass, in order to capture the effect of lamda in the linearized dynamics we have,
+
+        x_{k+1} = A x_k + B u_k + C lambda_k + d
+        y_k = G x_k + H u_k + J lambda_k + l
+        
+        Also, 0=< lamda_k has a complementarity constraint with y_k=>0
+
+Using this, we first zero out the columns in C and J where lamda_k is zero, then if lamda_k>0 then y_k =0 as
+per the constriant. Hence, 
+        0 = G x_k + H u_k + J lambda_k + l
+        lamda_k = -J^{+}(G x_k + H u_k + l)
+        lamda_k = A' x_k + B' u_k + l'
+
+Substituting this back in original dyanmics eqaution gives us A'' and B'' which is used in the backward pass
+
+        x_{k+1} = A'' x_k + B'' u_k + d''
+
+"""
+
+import numpy as np
+import pdb
+import matplotlib.pyplot as plt
+import pickle
+from toy_2d.src import vis_utils
+from toy_2d.src import file_utils
+from toy_2d.src.two_dim_polytope import TwoDimensionalPolytopeParams, \
+                                        TwoDimensionalPolytope
+from toy_2d.src.two_dim_system import TwoDimensionalSystemParams, \
+                                      TwoDimensionalSystem,TwoDSystemMagOnly, TwoDSystemForceOnly
+
+from toy_2d.src.two_dim_lcs_approximation import TwoDSystemLCSApproximation
+
+
+class iLQR():
+
+    def __init__(self, lcs, x_goal: np.ndarray, N: int, Q: np.ndarray, R: np.ndarray, Qf: np.ndarray):
+        self.lcs = lcs
+        self.x_goal = x_goal
+        self.u_goal = np.zeros((R.shape[0]))
+        self.N = N
+        self.dt = lcs.system.params.dt
+        self.Q = Q
+        self.R = R
+        self.Qf = Qf
+        # number of states
+        self.nx = Q.shape[0]
+        # number of actions
+        self.nu = R.shape[0]
+        # Solver parameters
```

Comment:
All of these solver parameters might be useful to package into one `@dataclass`-type class that can be included as an input to the `iLQR` class.